### PR TITLE
Reproducible build: introduce a tool to help reproducible build

### LINKED ISSUE #256: [Bug] td-shim should support reproducible build
There are some differences between 2 times build with same code base. Currently td-shim does not support reproducible build.

image

### PR DESCRIPTION
**Pull request by longlongyang**
fix #256

1. Introduce a reproducible tool, td-reproducible-tool, to help reproducible
2. Add strip = "symbols" to cargo.toml in root workspace to remove symbols

Signed-off-by: Longlong Yang <longlong.yang@intel.com>

### CODE REVIEW

**Comment by gaojiaqi7** [comment]
The CI failure caused by secure boot verification is due to `strip = "symbols"`, there may be some hidden optimizations which are not compatible with the new setting

File: td-reproducible-tool/readme.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+```
+1. Debug related sections (PE + ELF) 
+2. Code base path like \home\USERNAME\rust-td-payload\xxx\xxx.rs (PE + ELF) 
+3. CARGO_HOME like \home\USERNAME\.cargo\xxx\xxx.rs (PE + ELF) 
+4. RUSTUP_HOME like \home\USERNAME\.rustup\xxx\xxx.rs (PE + ELF) 
+5. TimeDateStamp field in COFF file header (PE only) 
+```
+
+Solution: <br>
+
```
Comment (jiangliu) [inline td-reproducible-tool/readme.md:16]: Could 2, 3, 4 also be solved by using build container to normalize the build env?

Comment (jyao1) [reply td-reproducible-tool/readme.md:16]: It is based upon the definition of "reproducible build".

In our definition, "reproducible build" means
1) Build with same code and version
2) Build with same compiler and version
3) Build at different time.
4) Build at different directory/path <==

File: td-reproducible-tool/readme.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+```
+1. Use strip = "symbols" option to [profile.release]
```
Comment (jyao1) [inline td-reproducible-tool/readme.md:18]: Please update to debuginfo

File: td-reproducible-tool/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    .iter()
+    .collect();
+
+    assert!(binary_path.exists());
+    println!("INFO: Found the compiled file: {:?}", binary_path);
+
+    // check out CARGO_HOME and RUSTUP_HOME
+    cargo_home = if cargo_home == "" {
+        if env::var("CARGO_HOME").is_err() {
+            panic!("Neither --cargo_home nor system environment for \"CARGO_HOME\" is found! ")
```
Comment (jyao1) [inline td-reproducible-tool/src/main.rs:86]: Can we make it optional?

Comment (jyao1) [inline td-reproducible-tool/src/main.rs:86]: Can we make it optional?

File: td-reproducible-tool/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        } else {
+            env::var("CARGO_HOME").unwrap()
+        }
+    } else {
+        cargo_home
+    };
+
+    rustup_home = if rustup_home == "" {
+        if env::var("RUSTUP_HOME").is_err() {
+            panic!("Neither --rustup_home nor system environment for \"RUSTUP_HOME\" is found! ")
```
Comment (jyao1) [inline td-reproducible-tool/src/main.rs:96]: Can we make it optional?

Comment (jyao1) [inline td-reproducible-tool/src/main.rs:96]: Can we make it optional?

Comment (longlongyang) [reply td-reproducible-tool/src/main.rs]: It is optional now, please check the latest code change.

Comment (longlongyang) [reply td-reproducible-tool/src/main.rs]: > Can we make it optional?

It is optional now, please check the latest code change.

File: td-reproducible-tool/src/main.rs (hunk 3 above)
Comment (longlongyang) [reply td-reproducible-tool/src/main.rs]: > Can we make it optional?

It is optional now, please check the latest code change.

Comment (longlongyang) [reply td-reproducible-tool/src/main.rs]: > Can we make it optional?

It is optional now, please check the latest code change.

File: td-reproducible-tool/readme.md (hunk 2 above)
Comment (longlongyang) [reply td-reproducible-tool/readme.md:18]: > Please update to debuginfo

Still using symbols in the latest code change.

File: td-reproducible-tool/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        } else {
+            env::var("CARGO_HOME").unwrap()
+        }
+    } else {
+        cargo_home
+    };
+
+    rustup_home = if rustup_home == "" {
+        if env::var("RUSTUP_HOME").is_err() {
+            panic!("Neither --rustup_home nor system environment for \"RUSTUP_HOME\" is found!\n")
```
Comment (jyao1) [inline td-reproducible-tool/src/main.rs:159]: still panic?

File: td-reproducible-tool/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        println!(
+            "INFO: -s or --strip_path is not specified, Skipping strip related rust file path."
+        );
+        return Ok(());
+    }
+
+    // Check out CARGO_HOME and RUSTUP_HOME, proceed to strip rust file path.
+    cargo_home = if cargo_home == "" {
+        if env::var("CARGO_HOME").is_err() {
+            panic!("Neither --cargo_home nor system environment for \"CARGO_HOME\" is found!\n")
```
Comment (jyao1) [inline td-reproducible-tool/src/main.rs:149]: still panic?

File: td-reproducible-tool/readme.md (hunk 1 above)
Comment (jyao1) [reply td-reproducible-tool/readme.md:16]: The final design: 
1 and 2 are handled by build option.
3 and 4 are handled by post-build tool as an option.
5. Timestamp is always cleared.

File: td-reproducible-tool/src/main.rs (hunk 5 above)
Comment (jyao1) [reply td-reproducible-tool/src/main.rs:159]: OK. I did not found -s or --strip_path in readme.

Please add it.

Comment (longlongyang) [reply td-reproducible-tool/src/main.rs:159]: > OK. I did not found -s or --strip_path in readme.
> 
> Please add it.

Sorry not to update readme. Now added, please check the latest readme.md

**Comment by liuw1** [comment]
I validated reproducible build with the latest code, there are still some differences. And the secure boot issue was found in the latest CI test.

image

**Comment by jyao1** [comment]
@liuw1 , please file a new issue, or reopen the old issue.

**Comment by longlongyang** [comment]
> I validated reproducible build with the latest code, there are still some differences. And the secure boot issue was found in the latest CI test.
> 
> image

This position is the where TimeDateStamp sit at, Did you run reproducible tool after every binary built?
And are you sure you observed the secure boot issue using the commit newer or equal to commit#fc436aabae?

**Comment by jyao1** [comment]
@longlongyang , I think we need update readme.md to add step for reproducible tool.
Otherwise, people may forget that step.

**Comment by longlongyang** [comment]
> @longlongyang , I think we need update readme.md to add step for reproducible tool. Otherwise, people may forget that step.

Okay, will update.

**Pull request merged**